tools/ion-bench/src/json/yy.h

```diff
@@ -0,0 +1,319 @@
+#pragma once
+#include <vector>
+#include <any>
+#include <optional>
+#include <utility>
+
+#include "../common.h"
+
+#include "yyjson.h"
+
+#include "common.h"
+
+namespace yyjson {
+
+class YYJson : public Library {
+   typedef json::JsonData<std::pair<yyjson_type, yyjson_subtype>> JsonData;
+   std::pair<yyjson_type, yyjson_subtype> make_type(yyjson_type tpe, yyjson_subtype sub) {
+      return std::make_pair(tpe, sub);
+   }
+   std::pair<yyjson_type, yyjson_subtype> make_type(yyjson_type tpe) {
+      return std::make_pair(tpe, YYJSON_SUBTYPE_NONE);
+   }
+
+   protected:
+      size_t _data_size;
+      std::vector<JsonData> _tape;
+
+   public:
+      const char *name() const override { return "yyjson"; }
+
+      static constexpr char const *suffix() {
+         return "json";
+      }
+
+      size_t data_size() const override {
+         return _data_size;
+      }
+
+      ValueStats deserialize(uint8_t *input, size_t in_size) override {
+         yyjson_read_err err;
+         auto doc = yyjson_read_opts((char *)input, in_size, 0, NULL, &err);
+         ValueStats stats = {0};
+
+         if (doc != nullptr) {
+            auto val = yyjson_doc_get_root(doc);
+            auto valstats = process_value(val);
+            stats += valstats;
+         } else {
+             printf("Error: %s at position %ld\n", err.msg, err.pos);
+         }
+
+         yyjson_doc_free(doc);
+         return stats;
+      }
+
+      ValueStats process_value(yyjson_val *val) {
+         ValueStats stats = {0};
+         switch (yyjson_get_type(val)) {
+            case YYJSON_TYPE_OBJ: {
+               stats.num_objs++;
+               auto objstats = process_object(val);
+               stats += objstats;
+               break;
+            }
+            case YYJSON_TYPE_STR: {
+               stats.num_strs++;
+               const char *s = yyjson_get_str(val);
+               std::string str(s, strlen(s));
+               stats.str_size += str.length();
+               break;
+            }
+            case YYJSON_TYPE_NUM: {
+               stats.num_nums++;
+               switch (yyjson_get_subtype(val)) {
+                  uint64_t i; int64_t j; double f;
+                  case YYJSON_SUBTYPE_REAL:
+                     f = yyjson_get_real(val);
+                     break;
+                  case YYJSON_SUBTYPE_UINT:
+                     i = yyjson_get_uint(val);
+                     break;
+                  case YYJSON_SUBTYPE_SINT:
+                     j = yyjson_get_sint(val);
+                     break;
+               }
+               break;
+            }
+            case YYJSON_TYPE_NULL:
+               stats.num_nulls++;
+               break;
+            case YYJSON_TYPE_ARR: {
+               stats.num_lists++;
+               auto valstats = process_list(val);
+               stats += valstats;
+               break;
+            }
+            case YYJSON_TYPE_BOOL:
+               stats.num_bools++;
```

Comment:
Should we retrieve the value of the boolean?

Reply:
Yea, I need to wire up a dependency on the bool so it doesn't get optimized out. For yyjson specifically getting a bool is a memory access, and a couple bit operations, so there aren't any side effects to keep the call around. I'll address it in the follow-up.